### PR TITLE
[WIP] Leibniz & Liskov part 2.

### PR DESCRIPTION
**Pull request by alexknvl**
## Preliminaries

```scala
type Void <: Nothing
type ¬[-A] = A => Void
type ⋁[+A, +B] = Either[A, B]
type ⋀[+A, +B] = Tuple2[A, B]
```

### [Inhabitance](https://github.com/scalaz/scalaz/pull/1534/files#diff-dfa9e61fe25a4177c0b164648882eb53R8)
`¬¬a` forms a monad.
```scala
type ¬¬[+A] = Inhabited[A]
final class Inhabited[+A](val run: (A => Void) => Void) { A =>
  def map[B](f: A => B): ¬¬[B] =
    new ¬¬[B](k => A.run(a => k(f(a))))

  def flatMap[B](f: A => ¬¬[B]) =
    new ¬¬[B](k => A.run(a => f(a).run(k)))
}
```

We can prove non-intuitionistic things under double negation such as [law of excluded middle](https://en.wikipedia.org/wiki/Law_of_excluded_middle)
```scala
object Inhabited {
  def witness[A](f: (A => Void) => Void): ¬¬[A] =
    new ¬¬[A](f)

  def lem[A]: ¬¬[¬[A] \/ A] =
    witness(k => k(-\/(a => k(\/-(a)))))
}
```

### [Proposition](https://github.com/scalaz/scalaz/pull/1534/files#diff-ccac84b8a45a02b434b8ef6e59467d36R9)
We'll say that type `a` is a [proposition](https://ncatlab.org/nlab/show/mere+proposition), if there is *at most* a single value of type `a` (currently not enforced), and there must be a `¬¬a ⟶ a`.

```scala
trait Proposition[A] {
  def prove(proof: ¬¬[A]): A
  // Not enforced at the moment!
  // def equal[B, C](implicit b: B <~< A, B: ¬¬[B], c: C <~< A, C: ¬¬[C]): B === C
}
```

`Proposition` forms an `IsoDivisible` (haven't added the typeclass yet):
```scala
trait IsoDivisible[F[_]]
  def isomap[A, B](fa: F[A])(f: Iso[A, B]): F[B]
  def zip[A, B](fa: F[A], fb: F[B]): F[(A, B)]
  def unit: F[Unit]
}
```
which makes it [easy](https://github.com/scalaz/scalaz/pull/1534/files#diff-99a90bce72fcbe2902f25c9ad2202d09R108) to derive new instances (e.g. a pair of propositions is a proposition).

## Type Poset

Types in a language with subtyping form a partially ordered set.

Let's define:

  * `a ~ b` - type equality
  * `a ≤ b` - subtyping relation

We can declare them in Scala as follows (see [this](https://github.com/scalaz/scalaz/blob/series/8.0.x/base/shared/src/main/scala/scalaz/data/Is.scala#L7) for an explanation):
```scala
// Leibnizian equality
trait ~[ A,  B] { def subst[F[_]] (fa: F[A]): F[B] }
trait ≤[-A, +B] { def subst[F[+_]](fa: F[A]): F[B] }

type ≠[A, B] = ¬[A ~ B]
```

Since they are inhabited by at most a single value, they are mere propositions and [we can](https://github.com/scalaz/scalaz/pull/1534/files#diff-dad13ef899b6f97285da2056afae8b64R216) [assert](https://github.com/scalaz/scalaz/pull/1534/files#diff-f1d9b24b3bdc455e0fff008a61940f4eR303):
```scala
def provedIs[A, B](f: ¬¬[A ~ B]): A ~ B = ...
def provedAs[A, B](f: ¬¬[A < B]): A < B = ...
```

Axioms for posets:

 * `a ≤ a` (reflexivity: every element is related to itself).
 * if `a ≤ b` and `b ≤ a`, then `a ~ b` (antisymmetry: two distinct elements cannot be related in both directions).
 * if `a ≤ b` and `b ≤ c`, then `a ≤ c` (transitivity: if a first element is related to a second element, and, in turn, that element is related to a third element, then the first element is related to the third element).

We hit our first stumbling block, anti-symmetry is not provable (at least in Scala2)! We will have to [assert it](https://github.com/scalaz/scalaz/pull/1534/files#diff-eeace4e8f23b2010316a803e776998d0R14):
```scala
def bracket[A, B](below: A < B, above: B < A): A ~ B = ...
```

Later on we will need:
  
  * `a ≸ b` - incomparable types
  * `a < b` - strict subtyping relation

```scala
type <[ A,  B] = (¬[A ~ B], A ≤ B)
type ≸[ A,  B] = (¬[A ≤ B], ¬[B ≤ A])
```

They should satisfy a number of poset axioms & consequences of those axioms:

 * `a < b ⟺ a ≤ b ⋀ ¬(a ~ b)` (by definition in Scala).
 * `a ≸ b ⟺ ¬(a ≤ b) ⋀ ¬(b ≤ a)` (by definition in Scala).
 * `a ≤ b ⟺ a < b ⋁  a ~ b` (can be proven in classical logic)
 * `¬(a ≤ b) ⟺ a ≸ b ⋁ b < a` (requires an additional axiom)
 * `¬(a ~ b) ⟺ a ≸ b ⋁ b < a ⋁ a < b` (requires an additional axiom)
 * ...

Only one of `a ~ b`, `a < b`, `a > b`, `a ≸ b` can be true for any given `a` and `b`. We can't prove it from inside the language, so we'll have to [assert it](https://github.com/scalaz/scalaz/pull/1534/files#diff-eeace4e8f23b2010316a803e776998d0R24):
```scala
def subtyping[A, B](ev: A ≠ B): ¬¬[(B < A) \/ (A < B) \/ (A ≸ B)] = ...
```

#### **Incomparability witnesses**
```
a ≸ b
⟺ ¬(a ~ b) ⋀ ¬(a < b) ⋀ ¬(b < a)
⟺ ¬(a ~ b ⋁ a < b ⋁ b < a)
⟺ ¬(a ≤ b ⋁ b ≤ a)
⟺ ¬(a ≤ b) ⋀ ¬(b ≤ a)

¬(a ≸ b)
⟺ ¬(¬(a ~ b) ⋀ ¬(a < b) ⋀ ¬(b < a))
⟺ a ~ b ⋁ a < b ⋁ b < a
⟺ a ≤ b ⋁ b ≤ b
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-f2ea869c68e653687175433d700a1170R23

#### **Strict Subtyping witnesses**
```
a < b
⟺ ¬(a ~ b) ⋀ a ≤ b

¬(a < b)
⟺ ¬(¬(a ~ b) ⋀ a ≤ b)
⟺ a ~ b ⋁ ¬(a ≤ b)
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-99a90bce72fcbe2902f25c9ad2202d09R96

### Type functions

Define "type functions" as type constructors and type lambdas and  write `f a` to mean type function `f` applied to type `a`.

Define:
  * `constant f` to mean `∀ a b. f a ~ f b` (constant function), this is also known as phantom variance.
  * `injective f` to mean `∀ a b. f a ~ f b ⟶ a ~ b` (injective function).
  * `covariant f` to mean `∀ a b. a ≤ b ⟶ f a ≤ f b` (monotonically increasing function).
  * `contravariant f` to mean `∀ a b. a ≤ b ⟶ f b ≤ f a` (monotonically decreasing function).
  * `strictly-covariant f` to mean `∀ a b. a < b ⟶ f a < f b` (strictly increasing function).
  * `strictly-contravariant f` to mean `∀ a b. a < b ⟶ f b < f a` (strictly decreasing function).
  * `invariant f` to mean `∀ a b. ¬(a ~ b) ⟶ f b ≸ f a`.

It is obvious that `constant f ⟹ covariant f` and `strictly-covariant f ⟹ covariant f` (analogously for contravariance). It is **not obvious at all** whether e.g. `strictly-covariant f ⟹ covariant f ⋀ ¬(constant f)`. We will need *parametricity axioms* (see below) to prove similar statements.

##### Lemma 1 (L1)
At most one of `strictly-covariant f`, `strictly-contravariant f`, `constant f`, `invariant f` can be true simultaneously. Proof is easy.

##### Lemma 2 (L2)
`strictly-covariant f ⟹ ¬(contravariant f)` and `strictly-contravariant f ⟹ ¬(covariant f)`

Proof:
WLOG we will only prove the first part. Suppose `f` is both strictly covariant and contravariant. Then for any `a` and `b` such that `a < b`, `f a < f b` and `f b ≤ f a ⟺ f b < f a ⋁ f b ~ f a`. Contradiction.

#### **Contrapositives**

```
strictly-covariant f
⟺ ∀ a b. a < b ⟶ f a < f b
⟺ ∀ a b. ¬(a < b) ⋁ f a < f b
⟺ ∀ a b. f a < f b ⋁ ¬(a < b)
⟺ ∀ a b. ¬(f a < f b) ⟶ ¬(a < b)
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-4108a2e8f554d2e45578ee61394e9581R37 (strict one is missing, oops)

#### **First parametricity axiom (P1)**
All type functions in Scala are either constant or injective.
https://github.com/scalaz/scalaz/pull/1534/files#diff-eeace4e8f23b2010316a803e776998d0R29

#### **Second parametricity axiom (P2)**
All injective type functions are either strictly covariant, strictly contravariant, or invariant.
https://github.com/scalaz/scalaz/pull/1534/files#diff-eeace4e8f23b2010316a803e776998d0R32

#### **Third parametricity axiom (P3)**
Applying an injective type function to a pair of incomparable types results in a pair of incomparable types.
```
∀ a b. a ≸ b ⟶ f a ~ f b ⋁ f a ≸ f b
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-eeace4e8f23b2010316a803e776998d0R19

#### **Injectivity & Constantness witnesses**
```
injective f ⟺ ¬(constant f)
⟺ ¬(∀ a b. f a ~ f b)
⟺ ∃ a b. ¬(f a ~ f b)

¬(injective f) ⟺ constant f
⟺ ¬(∀ a b. f a ~ f b ⟶ a ~ b)
⟺ ¬(∀ a b. ¬(f a ~ f b) ⋁ a ~ b)
⟺ ∃ a b. f a ~ f b ⋀ ¬(a ~ b)
⟺ ∃ a b. ¬(a ~ b) ⋀ f a ~ f b
```

#### **Variance bracket theorems**
```
covariant f ⋀ contravariant f 
⟺ (∀ a b. a ≤ b ⟶ f a ≤ f b) ⋀ (∀ a b. a ≤ b ⟶ f b ≤ f a)
⟺ ¬(0 ~ 1) ⋀ (f 0 ≤ f 1) ⋀ (f 1 ≤ f 0)
⟺ ¬(0 ~ 1) ⋀ (f 0 ~ f 1)
⟺ constant f
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-ab8787a369b05f5dafbb038a003fd759R55

```
covariant f ⋁ contravariant f
⟺ ¬(¬(covariant f) ⋀ ¬(contravariant f))
⟺ ¬(invariant f)
```

#### **Invariance witness**

```
invariant f
⟺ ¬(constant f) ⋀ ¬(strictly-covariant f) ⋀ ¬(strictly-contravariant f)
⟺ ¬(covariant f) ⋀ ¬(contravariant f)
⟺ ¬(∀ a b. a ≤ b ⟶ f a ≤ f b) ⋀ ¬(∀ a b. a ≤ b ⟶ f b ≤ f a)
⟺ ¬(∀ a b. ¬(a ≤ b) ⋁ f a ≤ f b) ⋀ ¬(∀ a b. ¬(a ≤ b) ⋁ f b ≤ f a)
⟺ (∃ a b. a ≤ b ⋀ ¬(f a ≤ f b)) ⋀ (∃ a b. a ≤ b ⋀ ¬(f b ≤ f a))
⟺ ∃ a b x y. a ≤ b ⋀ ¬(f a ≤ f b) ⋀ x ≤ y ⋀ ¬(f y ≤ f x)

¬(invariant f)
⟺ ¬(∀ a b. ¬(a ~ b) ⟶ f b ≸ f a)
⟺ ¬(∀ a b. a ~ b ⋁ f b ≸ f a)
⟺ ∃ a b. a ~ b ⋀ ¬(f b ≸ f a)
```

https://github.com/scalaz/scalaz/pull/1534/files#diff-0de33cfeb33e9db2baea7117b959c849R57
(incomplete)

#### **Covariance & Contravariance witnesses**
```
covariant f
⟺ ¬(strictly-contravariant f) ⋀ ¬(invariant f)
⟺ ¬(∀ a b. a < b ⟶ f b < f a) ⋀ ¬(∀ a b. ¬(a ~ b) ⟶ f a ≸ f b)
⟺ ¬(∀ a b. ¬(a < b) ⋁ f b < f a) ⋀ ¬(∀ a b. a ~ b ⋁ f a ≸ f b)
⟺ (∃ a b. a < b ⋀ ¬(f b < f a)) ⋀ (∃ a b. ¬(a ~ b) ⋀ ¬(f a ≸ f b))
⟸ ∃ a b. a < b ⋀ ¬(f b < f a) ⋀ ¬(a ~ b) ⋀ ¬(f a ≸ f b)
⟺ ∃ a b. a < b ⋀ ¬(f b < f a) ⋀ ¬(f a ≸ f b)
⟺ ∃ a b. a < b ⋀ (f a ≤ f b ⋁ f a ≸ f b) ⋀ ¬(f a ≸ f b)
⟺ ∃ a b. a < b ⋀ f a ≤ f b
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-79290fc972f9bcd2100a0232bacc2a60R62

W.L.O.G.
```
contravariant f
⟺ ∃ a b. a < b ⋀ f b ≤ f a
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-05d4b3ebf1faa70ef44a492218848670R55

#### **Strict Covariance & Contravariance witnesses**

```
strictly-covariant f ⟺ covariant f ⋀ injective f

¬(strictly-covariant f)
⟺ ¬(∀ a b. a < b ⟶ f a < f b)
⟺ ¬(∀ a b. ¬(a < b) ⋁ f a < f b)
⟺ ∃ a b. a < b ⋀ ¬(f a < f b)
```
https://github.com/scalaz/scalaz/pull/1534/files#diff-4108a2e8f554d2e45578ee61394e9581R67

### Putting it all together
```scala
trait Constant[F[_]] { 
	def apply[A, B]: F[A] ~ F[B]
}
trait Injective[F[_]] { 
	def apply[A, B](ev: F[A] ~ F[B]): A ~ B
}
trait Invariant[F[_]] {
	def apply[A, B](ev: F[A] ≠ F[B]): F[A] ≸ F[B]
}
trait Covariant[F[_]] { 
	def apply[A, B](ev: A ≤ B): F[A] ≤ F[B]
}
trait Contravariant[F[_]] {
	def apply[A, B](ev: A ≤ B): F[B] ≤ F[A]
}
trait StrictlyCovariant[F[_]] { 
	def apply[A, B](ev: A < B): F[A] < F[B]
}
trait StrictlyContravariant[F[_]] {
	def apply[A, B](ev: A < B): F[B] < F[A]
}

def bracket[A, B](below: A < B, above: B < A): A ~ B = ...

def incomparable[F[_], A, B](ncmp: A ≸ B): ¬¬[(F[A] ~ F[B]) ⋁ (F[A] ≸ F[B])] = ...

def subtyping[A, B](ev: A ≠ B): ¬¬[(B < A) \/ (A < B) \/ (A ≸ B)] = ...

def parametricity[F[_]]: ¬¬[IsConstant[F] \/ IsInjective[F]] = ...

def subtypeParametricity[F[_]](ev: IsInjective[F]): ¬¬[IsStrictlyContravariant[F] \/ IsStrictlyCovariant[F] \/ IsInvariant[F]] = ...
```

### CODE REVIEW

**Comment by tonymorris** [comment]
Cool 🥇

**Comment by TomasMikula** [comment]
Cool stuff here!

You assume this as an axiom (P1):

> All type functions in Scala are either constant or injective.

What about GADTs?

```scala
sealed trait Foo[A]
case class Bar(i: Int) extends Foo[Int]
```

`Foo` is a type function that assigns `Bar` to `Int` and is undefined for any other type (which we could interpret as assigning `Void` to every other type). So `Foo` is neither constant nor injective.

File: base/shared/src/main/scala/scalaz/data/Is.scala [hunk 1]
```diff
@@ -207,21 +208,11 @@ object Is {
...
+  def fromPredef[A, B](ev: A =:= B): A === B =
+    Axioms.predefEq(ev)
 
-  /**
-    * Unsafe coercion between types. `force` abuses `asInstanceOf` to
-    * explicitly coerce types. It is unsafe, but needed where Leibnizian
-    * equality isn't sufficient.
-    */
-  def unsafeForce[A, B]: A === B =
-    refl[Any].asInstanceOf[A === B]
```
Comment (TomasMikula) [inline base/shared/src/main/scala/scalaz/data/Is.scala]: I think we should keep `unsafeForce`, unless we can replace its uses with something else (like `proposition` below). All current uses of `unsafeForce` have been replaced with `asInstanceOf`, which is not an improvement.

File: base/shared/src/main/scala/scalaz/data/Iso.scala [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      case Right(i) => Right(ij.to(i))
+    }, {
+      case Left(b) => Left(ab.from(b))
+      case Right(j) => Right(ij.from(j))
+    })
+}
+object Iso {
+  def apply[A, B](implicit ab: Iso[A, B]): Iso[A, B] = ab
+
+  def unsafeTotal[A, B](ab: A => B, ba: B => A): Iso[A, B] = new Iso[A, B] {
```
Comment (TomasMikula) [inline base/shared/src/main/scala/scalaz/data/Iso.scala]: The "total" part of the name doesn't seem quite fitting.

**Comment by TomasMikula** [comment]
It doesn't seem quite fitting to call double negation `Inhabited`. I would suggest to call it `Consistent` (in the sense that (if inhabited) it would not lead to contradiction), or just leave it as `¬¬`.

`Proposition` with `def prove(proof: ¬¬[A]): A` might be better called `ProvableByContradiction`.

It would also be good to add some informal justification for the non-obvious axioms.

**Comment by jdegoes** [comment]
@alexknvl Excellent work. It would be good to include the preliminaries section in a README.md, just so we don't lose it.

File: base/shared/src/main/scala/scalaz/data/Inhabited.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  def lem[A]: ¬¬[¬[A] \/ A] =
+    witness(k => k(-\/(a => k(\/-(a)))))
+
+  def and[A, B](f: (A, B) => Void): ¬¬[¬[A] \/ ¬[B]] =
+    witness(p => p(\/-(b => p(-\/(a => f(a, b))))))
+
+  def imp[A, B](f: A => B): ¬¬[¬[A] \/ B] =
+    witness(k => k(-\/(a => k(\/-(f(a))))))
+
+  def pierce[A]: ¬¬[(¬[A] => A) => A] =
```
[the remark this answers is not in the document]
Comment (alexknvl) [reply base/shared/src/main/scala/scalaz/data/Inhabited.scala]:
```scala
type ¬[-A] = A => Nothing
type ¬¬[+A] = ¬[¬[A]]

def pierce[A]: ¬¬[((A => Nothing) => A) => A] =
  k => k((p: ¬[A] => A) => p((a: A) => k(_ => a)))
  
pierce[Int] : ¬¬[((Int => String) => Int) => Int]
```

You can upcast it to any `((A => B) => A) => A`.

[the remark this answers is not in the document]
Comment (TomasMikula) [reply base/shared/src/main/scala/scalaz/data/Inhabited.scala]: 👍

**Comment by jdegoes** [comment]
@TomasMikula I prefer the names `Consistent` and `ProvableByContradiction` 👍

**Comment by TomasMikula** [comment]
I realized that `Consistent` is not a great name, either. While it is true that if we can derive ¬¬A then A is consistent, there are consistent propositions A for which we don't have ¬¬A. Maybe just call it `NotNot`.

**Comment by jdegoes** [comment]
Maybe NotInconsistent?

**Comment by TomasMikula** [comment]
"Not inconsistent" is the same as "consistent". Consistency is a term from meta-theory where we happily apply rules of classical logic.

**Comment by alexknvl** [comment]
Any type `A` for which you can write a total `(A => Void) => Void` is inhabited by some value and vice-versa, hence the name. It's a sufficient and necessary condition AFAICT.

https://en.wikipedia.org/wiki/Type_inhabitation
https://ncatlab.org/nlab/show/inhabited+set

> Finally, we note that the propositions-as-types correspondence can be viewed in reverse, al-
lowing us to regard any type `A` as a proposition, which we prove by exhibiting an element of `A`. Sometimes we will state this proposition as “`A` is inhabited”. That is, when we say that `A` is inhabited, we mean that we have given a (particular) element of `A`, but that we are choosing not to give a name to that element. Similarly, to say that `A` is not inhabited is the same as to give an element of `¬A`.

> The propositional truncation, also called the (−1)-truncation, bracket type, or squash type, is an additional type former which “squashes” or “truncates” a type down to a mere proposition, forgetting all information contained in inhabitants of that type other than their existence.

>  Similarly,  we will say that a type `A` is merely inhabited to mean that its propositional truncation `‖A‖` is inhabited (i.e. that we have an unnamed element of it).

https://hott.github.io/book/nightly/hott-online-1075-g3c53219.pdf

I am not entirely sure, but I think that for most intents and purposes `¬¬A` is the same as "propositional truncation", which indicates whether a type is inhabited.

**Comment by alexknvl** [comment]
This might also be relevant

http://eprints.nottingham.ac.uk/38092/1/1610.03346v1.pdf

> If we regard a type as the correspondent of a mathematical statement, a proposition, and its inhabitants of proofs thereof, we have to notice that there is a slightly unsatisfactory aspect.  A proof of a proposition in mathematics is usually not thought to contain any information apart from the fact that the proposition is true; however, a type can have any number of inhabitants, and therefore any number of witnesses of its truth.  Hence it seems natural to regard only some types as propositions, namely those which have at most one inhabitant.  The notion of propositional truncation assigns to a type the proposition that  this  type  is  inhabited.   To  make  the  connection  clearer,  these  types  are  even  called propositions, or h-propositions, in homotopy type theory.  With this in mind, we want to be able to say that a type is inhabited without having to reveal an inhabitant explicitly.  This is exactly what propositional truncation `‖−‖ : U → U` (where we write `U` for the universe of types) makes possible.  On the other hand, should `A` have only one inhabitant up to the internal  equality,  this  inhabitant  can  be  constructed from  an  inhabitant  of `‖A‖`.   This  is a crucial difference between propositional truncation and double negation.

>  [...] the double negation `¬¬A`, another notion of existence which, however, is often not useful as it generally only allows to prove negative statements.

> In  a  constructive  setting,  the  question  how  to  express  that  “there  exists  something” in a type `X` is very subtle.  One possibility is to ask for an inhabitant of `X`, but in many cases, this is too strong to be fulfilled.  A second possibility, which corresponds to our above definition of separated, is to ask for a proof of `¬¬X`.  Then again, this is very weak, and often too weak, as one can in general only prove negative statements from double-negated assumptions.

> This  fact  has  inspired  the  introduction  of squash  types,  and  similar, bracket types. These lie in between of the two extremes mentioned above. [...] For any type `X`, we postulate that there is a type `‖X‖` that is a proposition, representing the statement that `X` is inhabited.  The rules are that if we have a proof of `X`, we can, of course, get a proof of `‖X‖`, and from `‖X‖`, we can conclude the same statements as we can conclude from `X`, but only if the actual representative of `X` does not matter.

**Comment by TomasMikula** [comment]
Thanks for quoting relevant literature! These excerpts, though, seem to distinguish between double negation and (mere) inhabitance.

**Comment by alexknvl** [comment]
This is a more direct translation from the paper above.

```scala
import scala.language.higherKinds

type ![A] = Bracketed[A]
type ===[A, B] = Is[A, B]
type <=[-A, +B] = As[A, B]
type <[-A, +B] = StrictAs[A, B]

sealed abstract class Is[A, B] {
  def subst[F[_]](fa: F[A]): F[B]
}
object Is {
  final class Refl[A] extends Is[A, A] {
    def subst[F[_]](fa: F[A]): F[A] = fa
  }
  
  def force[A, B] = (new Refl[A]).asInstanceOf[Is[A, B]]
}

sealed abstract class As[-A, +B] {
  def subst[F[+_]](fa: F[A]): F[B]
}
object As {
  final class Refl[A] extends As[A, A] {
    def subst[F[+_]](fa: F[A]): F[A] = fa
  }
}

sealed abstract class StrictAs[-A, +B] {
  def notEqual[AA <: A, BB >: B]: (AA === BB) => Void
  def conforms: A <= B
}

type Void <: Nothing
trait Bracketed[+A] {
  def apply[P](f: A => P)(implicit P: IsProposition[P]): P
}
object Bracketed {
  def pure[A](a: A): Bracketed[A] = new Bracketed[A] {
    def apply[P](f: A => P)(implicit P: IsProposition[P]): P = f(a)
  }
  
  implicit def prop[A]: IsProposition[![A]] = new IsProposition[![A]] {
    def apply[X, Y](implicit X: SingletonOf[X, ![A]], Y: SingletonOf[Y, ![A]]): X === Y =
    	Is.force[X, Y] // I doubt it is possible to prove it. 
  }
}


trait SingletonOf[A, +B] {
  def conforms: A < B
  def bracketed: ![A]
}

trait IsProposition[A] {
  def apply[X, Y](implicit X: SingletonOf[X, A], Y: SingletonOf[Y, A]): X === Y
}
```

**Comment by TomasMikula** [comment]
With this representation, I don't see how a user could prove `Bracketed[A]` without first proving `A`.

What also confuses me is the cyclic definition: `Bracketed` depends on `IsProposition` depends on `SingletonOf` depends on `Bracketed`.

Anyway, I was not asking for a change of representation, just that we use more precise names.

**Comment by alexknvl** [comment]
Closed to rework & split into smaller PRs.

**Pull request closed without merging**